Spack Integration Testing

johnwparent opened the pull request:
# Integration testing is now! 

This PR introduces integration testing to the Spack repository in the form of a new test repo and a new Github actions workflow.

The new testing repo contains "integration test packages" that are effectively normal Spack packages, but have sufficiently simple sources and build systems (if any) that executing the build should not introduce a vector for test failure, and instead have package recipes that are crafted to test specific workflows within Spack. 

This ensures each of Spack's components can integrate successfully and with expected behavior in production like scenarios.

The Github actions workflow processes the packages available in the integration test repo, and creates a child job installing and testing each package on each of Windows, Ubuntu, and MacOS. 

## Motivation

Spack has gaps in its testing. 

Reading the above may draw parallels to what we currently do in Gitlab, build actual packages in a production environment to ensure changes introduced by a given PR don't break our capacity to continue building those packages. However, this has a few limitations. 
1. It does not run on changes to Spack core. If a PR were to break, say staging, by changing how stages are established in core. Gitlab would run no piplines, and we would allow this bug to merge (if uncaught by unit tests, which does happen). This disconnect is a large vector for introducing bugs into Spack.
2. It tests very specific packages. A change that does happen to trigger pipelines will still only test the behavior that just happens to be executed by the packages we are choosing to test. This is due to our package first testing approach in Gitlab CI, which while crucial for Spack to maintain our ability to build critical software, does not make for a great Spack core testing strategy. 
3. Real packages introduce failing points that are unrelated to  Spack core and are thus unsuited for testing Spack core. Time must be spent ferreting out whether a failure is a core or package issue.
4. It's very expensive. Adding more packages as "integration tests" will still have all the issues above, and cost more in AWS fees.

Spack's unit tests additionally do a good job of verifying the behavior of the units we test, but have the limitations of not testing facets of Spack in conjunction with each other (or at least they shouldn't be if they're true unit tests), and are limited to the test authors ability to predict possible edge cases and points of failure which can be prone to missing cases or having overly contrived test cases that do not evaluate the full spectrum of input/context that may be present in a more complex ecosystem.

Integration testing special packages in Github actions has a few advantages:

1. It allows us to have package builds on changes to Spack core, enabling consistent integration testing at the points where Spack is most likely to break large scale behavior. 
2. It's less contrived and more robust than simple unit testing, gives us a picture of how Spack is functioning as a whole.
3. It can save AWS costs. In the event a PR that changes core lands, but breaks core in a way uncaught by unit tests (no Gitlab CI on pure core changes), Gitlab's CI will be run on all package PRs, but will fail due to the core breakage. Further, to catch spurious system/runner failures, each job will be retried 3 times. This results in potentially numerous jobs being run that are destined to fail, wasting AWS credits, resources, and energy (both human and electrical). With integration testing, we can catch those failures before they land in develop, and reduce our wasted AWS credits on rebuilds destined to fail.
4. It allows for testing specific behavior. Gitlab's integration tests only test specific packages, they're not oriented toward testing specific Spack behaviors. With the integration testing repo, we can develop package recipes and packages that exercise specific Spack workflows. 


To do (future work):

- [ ] Introduce mechanisms for skipping certain packages based on a set of conditions (platform, available compilers, etc) 
- [ ] Introduce X-fail behavior for packages we expect to fail
- [ ] Expand number and coverage of tests
- [ ] Determine which "changes" should dispatch this workflow, i.e. better integration into Spack's current GHA based testing infrastructure


See [this doc](https://docs.google.com/document/d/16PttUheznvs38jPM30ln5Viz-3qg00bRmSwsM_Kdh-w/edit?tab=t.0) for significantly more detail. 
(please contact @johnwparent if you have any issues accessing the doc and would like to be able to)


### Bug Fix:

This PR revealed a bug in Windows nested logging (testing logging within the build log context produced this), when entering the nested logging context, the current stdout/err, which were redirected for the outer log context, were redirected again, and then closed on exit of the inner context, meaning when the outer context went to refer to these fds, they were invalid, and Spack crashed.

This PR also caught #51564

johnwparent commented:
cc @tldahlgren @zackgalbreath @kwryankrattiger @becker33

johnwparent commented:
Ensure what's going on with RTD

johnwparent commented:
Once there's consensus I'll squash and do the DCO correctly

johnwparent commented:
Failing Windows check is expected, integration tests have already caught a bug!

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/smoke_package.template:
```diff
@@ -0,0 +1,25 @@
+# Copyright Spack Project Developers. See COPYRIGHT file for details.
```
Until it is exercised, I worry that there will be significant churn for any template files added (or they may go on unused). I think ultimately the span of tests generated for integration testing will stress the usefulness of any template-based approach to "summarizing" packages.

johnwparent replied:
Good point, removed.

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/packages/smoke1/package.py:
```diff
@@ -0,0 +1,35 @@
+    """
+
+    homepage = "https://spack.io"
+    url = "file:" + pathname2url(os.path.join(os.path.dirname(__file__), "smoke1-0.1.tgz"))
```
It would be nice if we could generate the tar.gz archives on-demand vs. distributing them with the source code (that also requires defining the hash). `__file__` is accessible here, and you could write to `/the-package-basedir/package-src.tar.gz` if it doesn't exist yet (which I think means it would persist across separate cmdline invocations of spack).

johnwparent replied:
Would this be executed from the file? That feels like it could explode pretty quickly if we're doing this for N files. I also worry that generating this dynamically at test time leaves more room for errors unrelated to genuine regressions.
We also only do one command line run of Spack per test package in CI, so its useful for local unit tests but not so much GHA.

scheibelp replied:
> Would this be executed from the file?

yes

> That feels like it could explode pretty quickly if we're doing this for N files.

Not sure what is meant by this: if all the `spack install` invocations run in the same instance, each package.py would generate its associated `.tar.gz` once 

> generating this dynamically at test time leaves more room for errors unrelated to genuine regressions.

It might be worth manually maintaining a .tar.gz for this first package for this reason, but I think it's going to be annoying to run tests locally (every tweak to the source code means (a) new tarbal (b) new checksum).

> We also only do one command line run of Spack per test package in CI, so its useful for local unit tests but not so much GHA.

Do all the smoke test installs occur in the same instance? If so: then storing the tarball in the package directory should avoid repeated calculations. If not, then I could see it being problematic to regenerate them in the context of each `spack install`.

johnwparent replied:
> Do all the smoke test installs occur in the same instance?

No, each `spack install` command requires a fresh clone of spack as they are each independent jobs.

> It might be worth manually maintaining a .tar.gz for this first package for this reason, but I think it's going to be annoying to run tests locally (every tweak to the source code means (a) new tarbal (b) new checksum).

That's a good point, it would be very tedious to update. Maybe we can include a script to help with this, that way we avoid the repeated tarball creations and also avoid the tedium of continuously having to recompute these?

scheibelp replied:
> command requires a fresh clone of spack as they are each independent jobs.

Is there a way to run them together? Even ignoring the tarball issue, recloning spack for each individual package install could draw out the amount of time this check takes.

> Maybe we can include a script to help with this, t

A script would be good, although a secondary concern is avoiding storing binary objects in the git repository where possible (there's a few places we do it because we are testing decompression etc. and I think it's unavoidable there, but here I'm not sure)

johnwparent replied:
> Is there a way to run them together? Even ignoring the tarball issue, recloning spack for each individual package install could draw out the amount of time this check takes.

Running each build in isolation allows for a.) parallelization of each of the install tests, and b.) I consider the isolation to be a positive feature of the testing, to allow us to test the specific behavior indicated by the particular package without side effects.

> A script would be good, although a secondary concern is avoiding storing binary objects in the git repository where possible (there's a few places we do it because we are testing decompression etc. and I think it's unavoidable there, but here I'm not sure)

Is it a security concern, or a space concern?

 I'm primarily worried dynamically constructing the tarballs may be error prone and lead to somewhat confusing/hard to diagnose errors, particularly if it's failing for a reason other than what the particular integration test is attempting to validate.

scheibelp replied:
> I consider the isolation to be a positive feature of the testing,

I agree this is a benefit, but I consider the cost too high. This check is not lightweight if we accumulate 10-20 smoke test packages. To be clear, regardless of whether tarballs are generated dynamically, unless the tests in isolation can make use of the same spack instance vs. cloning it each time, simply that is a significant cost.

> Is it a security concern, or a space concern?

space concern: each time tarball changes, it replicates in the git history. Even if the tarball is relatively small like 1kb, a series of changes could increase the size over time (and when that happens, there's not much to be done to undo it).

> I'm primarily worried dynamically constructing the tarballs may be error prone and lead to somewhat confusing/hard to diagnose errors

Such functionality should be tested, I agree it will lead to debugging related to the feature itself rather than the behaviors the smoke tests are attempting to check. I consider that a cost worth paying.

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,80 @@
+Tests production context runs of Spack
+
+* Isolates behavior from individual package quirks
+* Guaruntees builds without worrying about Gitlab CI's pruning mechanisms (no "nothing to rebuild")
```
(minor) typo -> Guarantees

johnwparent replied:
There's a bunch of these 😬 Thanks

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,80 @@
+
+* Isolates behavior from individual package quirks
+* Guaruntees builds without worrying about Gitlab CI's pruning mechanisms (no "nothing to rebuild")
+* Directly tests specific Spack behaviors via controlled, deliberatly composed packages relying on specific behaviors
```
(minor) awkward wording: "specific" twice. I think the key feature is that we generate the source code ourselves, so we can exercise any behavior in Spack that we want (I didn't see any of the points mention source code).

johnwparent replied:
Great suggestion, thanks

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,80 @@
+
+Installing - Installing the package from the build tree
+
+Testing - Validates the software built and installed actually runs
```
(minor) I recommend bulleted list

Also, I think this item "testing" should be more-specifically worded: we are testing the ability to define tests inside of our packages and run them to verify a correct installation.

Overall, I think for all `x - y`, we should drop the `x` or put all `y` into sub-items (for now for example we only exercise one type of fetching, but we could add a git repo and test that too.

johnwparent replied:
Great suggestion, rephrased.

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,80 @@
+
+#### MPI
+
+MPI introduces a lot Spack logic. With a properly constructed integration test, we can have guarantees
```
(minor) grammar -> a lot of

scheibelp commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,105 @@
+        run: |
+          . share/spack/setup-env.sh
+          packages_list="$(spack list --repo integration_test | python3 -c "import sys; print([f'"{x}"' for x in sys.stdin.read().split('\n') if x])")"
+          echo "packages=${packages_list}" >> "$GITHUB_OUTPUT"
```
I recommend adding a TODO: there might be packages that we're only interested in as dependencies of other packages. The way concretization works, this could end up installing `O(n^2)` packages for a repo of size `N`: you get effects like this for example if for `x->y`, `x` has `depends_on(y~default_on)`; `install x` and `install y` will generate 3 specs.

Overall, I think that adding new packages to the integration test repo will be rare enough that it could be fine to list them out explicitly.

johnwparent replied:
Agreed, added the TODO

haampie replied:
Just write a spack.yaml...

kwryankrattiger replied:
My gripe here was that it would be nice to build each integration test package as it's own job in a matrix rather than as a batch. That way the tests that are failing would be more evident vs having to parse longer logs of output.

If we are going to do a batch of packages, I agree with @haampie that it should just be a `spack.yaml`, it makes it easier to enable/disable integration tests that may break.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -812,6 +818,7 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             self._thread.join()
             self.stdout.close()
             self.stderr.close()
+            self.echo_writer.close()
```
Many changes in this file look like refactoring except for this - did I get that right? Can you add a short description of fixed bug to the PR description?

johnwparent replied:
The getattr calls are meaningful as well, updating the description.

scheibelp reviewed (changes requested):
A few doc requests and a couple code change requests. The most-potentially involved is generating `.tar.gz`'s and hashes on demand - I have a suggestion for how to potentially do that though, which would avoid having to manually generate/update tar.gz and hashes for all the integration test packages.

johnwparent committed:
Overhaul initial integration test CI and Repo

CI checks are more efficient

Repo logic dynamically generates tarballs, improves docs, and notes
future work

* Add todo for case of dependency only build integration packages

* dynamically generate tarball

* Correct relative path inside generated tarball

* set fetch_depth=1

* Correct dynamically generated tarball structure

* tarball creation method takes versions

* integration readme: fix typo, remove cruft

* smoke test package: hash file in chunks

* Ensure repo configs are not shared accross spacks

* Dont bootstrap clingo

* Gives jobs names

* Move flow to prefer local env file

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Overhaul initial integration test CI and Repo

CI checks are more efficient

Repo logic dynamically generates tarballs, improves docs, and notes
future work

* Add todo for case of dependency only build integration packages

* dynamically generate tarball

* Correct relative path inside generated tarball

* set fetch_depth=1

* Correct dynamically generated tarball structure

* tarball creation method takes versions

* integration readme: fix typo, remove cruft

* smoke test package: hash file in chunks

* Ensure repo configs are not shared accross spacks

* Dont bootstrap clingo

* Gives jobs names

* Move flow to prefer local env file

Signed-off-by: John Parent <john.parent@kitware.com>

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,81 @@
+
+##### Challenges
+
+Feasabiliity to be investigated\
```
(minor) typo

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## FAQ
+
+Coming Soon
```
Will this be in before the PR merges? I think the section can disappear entirely if not.

scheibelp commented on share/spack/setup-env.sh:
```diff
@@ -284,6 +284,7 @@ else
     _sp_share_dir="$(cd "$(dirname $_sp_source_file)" > /dev/null && pwd)"
     _sp_prefix="$(cd "$(dirname $(dirname $_sp_share_dir))" > /dev/null && pwd)"
 fi
+
```
Vestigial space? I assume this is based on some edit that was later reverted.

johnwparent replied:
Yes, good catch.

scheibelp commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,113 @@
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-13, macos-14]
```
I think we are up to mac-os 15 now, so maybe replace the macos ones with `macos-latest`.

johnwparent replied:
Updated to same version range as CI checks, latest Mac image with intel support + macos-latest (which is arm)

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/packages/smoke1/package.py:
```diff
@@ -0,0 +1,55 @@
+        with tarfile.open(archive, "w|gz") as tgz:
+            tgz.add(os.path.join(package_root, "smoke1-src"), arcname=".")
+    sha256_hash = hashlib.sha256()
+    # TODO: maybe chunk this?
```
(question) would `hashlib.file_digest` help?

johnwparent replied:
My plan was to use `hashlib.update(chunked_data)` and then `hexdigest`, I thought `file_digest` was more recent than our oldest supported python, but yes, if we can use it, it would be better.

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -668,42 +668,48 @@ def __init__(self, sys_attr):
             self.libc.fflush(None)
         else:
             self.libc.fflush(self.c_stream)
-        self.sys_stream.flush()
```
(question) It's difficult for me to see how this changes the logic: e.g. originally `self.sys_stream` was set to be `getattr(sys, self.sys_attr)` (after we assigned that to the new stream, at the end of `redirect_stream`), so I would think the before/after lines here have the same effect. I feel similarly about all changes in this file above this point (i.e. they look like they do not change the logic).

scheibelp replied:
We spoke more about this today. My attempt at an explanation for this (why it helps with nested calls).

IMO it's useful to think about this for stdout (this method is called for stderr as well, but keeping it to one allows ignoring some abstractions):

redirect_stream for stdout

- stores sys_stream = stdout (the file object)
- the nested invocation calls `close` on stdout (the file object)
- sys.stdout is reassigned to something valid in the nested invocation
- But sys_stream is storing the original (now invalid) file object

scheibelp commented on lib/spack/spack/llnl/util/tty/log.py:
```diff
@@ -668,42 +668,48 @@ def __init__(self, sys_attr):
-        # Set sys_stream to a new stream that points to the redirected fd
-        new_buffer = open(self.orig_stream_fd, "wb")
+        os.dup2(to_fd, original_fd)
+        # Set self.sys_stream to a new stream that points to the redirected fd
```
`self.sys_stream` no longer exists

I would recommend something like `set the std io handle (out/err) to point to to_fd`

johnwparent commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,114 @@
+          fetch-depth: 1
+      - name: "Add Integration test repo"
+        run: |
+          mkdir ~/.spack
```
Feels like this shouldn't be necessary

scheibelp replied:
I think this is resolved by https://github.com/spack/spack/pull/51700

scheibelp commented on var/spack/test_repos/spack_repo/integration_test/packages/smoke1/package.py:
```diff
@@ -0,0 +1,58 @@
+        with tarfile.open(archive, "w|gz") as tgz:
+            tgz.add(os.path.join(package_root, "smoke1-src"), arcname=".")
+    sha256_hash = hashlib.sha256()
+    # TODO: maybe chunk this?
```
I think you have now done this TODO (so the comment can be removed)

haampie commented on var/spack/test_repos/spack_repo/integration_test/packages/smoke1/package.py:
```diff
@@ -0,0 +1,58 @@
+
+    license("MIT")
+
+    version("0.1", sha256=make_pkg_tarball("0.1"))
```
This is rather bizarre. I don't want this code to run during import. It's not even process safe.

I would suggest not to include these sources in Spack but use a separate git repo.

scheibelp replied:
I think the sources should live in the spack repo, or else they will be a pain to update

I'm not sure what you mean by not process safe - like if two spack instances simultaneously perform operations with the smoke test repo? I don't think we need that anywhere. Would a warning be sufficient?

Perhaps I made the wrong suggestion and we can disable checksums entirely for this internally-managed source (and therefore wouldn't need to calculate them).

scheibelp replied:
With a minor change to spack, packages can declare 

```
version(1.0, local="1.0")
```

and this fetcher will copy the `1.0` directory under ths spack package dir. I think this would resolve both our concerns.

haampie replied:
> I think the sources should live in the spack repo, or else they will be a pain to update

I disagree. Just look at these sources, it's "hello world".

> With a minor change to spack

We already have `dev_path=...`, can't you use that if you insist?

scheibelp replied:
I want adding a new smoke test to be as easy as possible. When someone wants to create a new package/source I think it will be annoying to create a new git repo for it (regardless of how much or how little it contains).

(`spack develop` could be useful for changing existing tests, but not for adding new smoke test packages)

scheibelp replied:
`share/spack/qa/develop_env_specs.py` was added which automatically develops all root specs in an environment (as a matter of convention, assumes that the source is located in the package repository directory under `{pkgname}-src`), the CI uses this on the `integration_env` environment. So I think a compromise has been reached here.

haampie commented on lib/spack/spack/llnl/util/tty/log.py:
Please extract to a separate PR.

scheibelp replied:
I think the integration check added in this PR will fail on windows without these changes. I think it makes sense to include these changes, perhaps as a distinct commit (and then rebasing vs. squashing).

haampie replied:
Just open a separate PR, it's pretty standard. It should be reviewed on its own, and possibly backported and labelled as such. This PR has been open half a year, so I don't get the rush.

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,132 @@
+      - name: "Spack install"
+        id: install
+        run: |
+          pip install --upgrade pip clingo
```
Why pip?

scheibelp replied:
Without this, clingo was being installed by spack in each of the runners without a binary cache - this is faster for the checks.

haampie replied:
No, it's Linux and macOS for which we have optimized binaries. `pip install clingo` gives you a different clingo version with binaries that are > 2x slower. In fact that are no wheels for Python 3.14 would it become a default.

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,132 @@
+        run: |
+          chmod +x bin/spack
+          . share/spack/setup-env.sh
+          git clone --depth=1 https://github.com/spack/spack-packages.git spackages
```
You could use the checkout action here.

johnwparent replied:
Yeah might as well.

haampie commented on var/spack/test_repos/spack_repo/integration_test/packages/smoke1/smoke1-src/add.cxx:
```diff
@@ -0,0 +1,12 @@
+        std::cerr << "Improper number of arguments (" << argc-1 << "), 2 were expected\n";
+        return 1;
+    }
+    int a = atoi(argv[1]);
```
This is not C++, this is C. If you want to write C++, then import the [relevant header](https://en.cppreference.com/w/cpp/string/byte/atoi.html)

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,132 @@
+        uses: actions/upload-artifact@v4
+        with:
+          name: spack-checkout
+          path: .
```
This one puzzles me. Do you really need GitHub actions to upload the entire Spack checkout? What's the goal?

johnwparent replied:
That was a request based on trying to speed up this CI check by avoiding an additional clone. Since I'm removing the scan step, this will go as well.

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,132 @@
+        id: scan
+        run: |
+          . share/spack/setup-env.sh
+          packages_list="$(spack list --repo integration_test | python3 -c "import sys; print([f'"{x}"' for x in sys.stdin.read().split('\n') if x])")"
```
What's wrong with `spack list --repo integration_test >> "$GITHUB_OUTPUT"`? Does it have to be so dynamic? I feel like a simple, declarative `spack.yaml` would be desirable instead of all these commands, that makes local troubleshooting much easier. We do that elsewhere.

johnwparent replied:
Originally each integration test package was tested in its own dynamically generated child job (one per windows/linux, 2 for macos), but since we no longer do that, this `scan` job is, I think, completely unnecessary, and you're right, we can just commit a environment yaml and use that.

kwryankrattiger commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,76 @@
+  * Run them to verify a correct installation
+
+Each of these higher level tasks exercises many more atomic sections of Spack that are well covered by our unit tests.
+The purpose served by this test is to validate the integration and compatibility of each of these more atomic components (hence, integration testing :) )
```
```suggestion
The purpose served by these tests is to validate the integration and compatibility of each of these more atomic components (hence, integration testing :) ).
```

kwryankrattiger commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,76 @@
+
+### Roadmap for future Tests
+
+Build systems - Expand beyond CMake
```
```suggestion
The following road map for integration testing outlines several key areas to enhance the testing framework. These include expanding beyond CMake build systems, developing  system configuration specific tests, and more complex repository and package configurations.

The integration testing road map:

1. **Build Systems**
   - Autotools
   - Meson
   - Python
   - Etc.

2. **System specific tests**
   - Platform
   - Architecture
   - Operating System

3. **Configs**
   - Different configuration setups

4. **Repos**
   - Multi-package repository repos
   - Mixed package API versions
   - Etc.

5. **Packages**
   - Different types of source URLs
   - Packages with more complex configuration
   - Packages with dependencies
   - Virtual packages
   - Etc.

6. **Libs and Executables**
   - Test all types

7. **Modules**
   - Test usage of modules
```

kwryankrattiger commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,76 @@
+
+Libs and executables - test all types
+
+Modules - test usage of modules
```
What kind of modules? C++ modules, python modules, spack modules...

kwryankrattiger commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,76 @@
+
+Packages - Packages doing more complex things, packages that depend on each other, virtuals, etc
+
+Libs and executables - test all types
```
Which types?

kwryankrattiger commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,76 @@
+#### MPI
+
+MPI introduces a lot of Spack logic. With a properly constructed integration test, we can have guarantees
+of basic MPI related behaviors functioning in Spack and behaving as expected without an expensive or potentiall no-op Gitlab CI run.
```
Keep on same line as above

kwryankrattiger commented on var/spack/test_repos/spack_repo/integration_test/README.md:
```diff
@@ -0,0 +1,76 @@
+
+##### Challenges
+
+CUDA is available on GHA runners, but GPUs are not universally available and may cost money/have a higher overhead. We may need to self host, reach out to Github about pricing options, or simply be incredbly selective about when this check is run
```
```suggestion
There is some CUDA availability from GitHub Actions runners; however, GPUs are not universally available. Additional investigation is required to determine if and how integration tests for GPU should be run.
```

kwryankrattiger reviewed (commented):
Mostly documentation nits

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,106 @@
+  cancel-in-progress: true
+
+jobs:
+  integration-nix:
```
nitpick: `nix` is ambiguous and rather odd in the context of Spack the package manager.  I would recommend to avoid that term. Especially since you're only testing Linux and macOS there's no point in talking about "*nix" if that was the intention.

johnwparent replied:
Fair point, will re-name.

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,106 @@
+
+jobs:
+  integration-nix:
+    name: Integration-Build-${{ matrix.os }}
```
nitpick: other actions don't use uppercase like this (Power-Shell-Style?). I would make it consistent with the rest, otherwise it stands out more than it needs.

johnwparent replied:
Good point, thanks.

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,106 @@
+            path: spackages
+      - name: Add builtin with shallow clone
+        run: |
+          chmod +x bin/spack
```
```suggestion
```

haampie commented on .github/workflows/integration.yml:
```diff
@@ -0,0 +1,106 @@
+          name: "broken-${{ matrix.os }}-stage"
+          path: $FAILURE_STAGE
+
+  integration-win:
```
The only difference between this and the above is shell setup. Feels like it should be possible to combine this by making setup-env part of the shell on linux/macos? Alternatively, don't use shell setup but just run the Spack executable

That avoids that the two diverge, which they already do regarding `fetch-depth`.

johnwparent replied:
Since spack built clingo is ~2.2x faster than pip installed clingo, and the BC hit on not Windows is fast, we should be allowing Spack to bootstrap it for Linux and Mac, but for since we're not doing many concretizations, and it takes about ~6 minutes to build on Windows (no BC), should they still diverge where we want to pip install clingo on Windows?

If you think we should bootstrap clingo everywhere, happy to make efforts to consolidate as suggested.

haampie replied:
You can do a step `run: pip install clingo` with `if: ${{ matrix.os }} == ...` if that's the only difference

johnwparent committed:
Integration Testing: initialize integration test repo

* add repo infrastructure for smoke testing
* add a simple package for basic integration smoke testing

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Integration Testing: initialize integration test repo

* add repo infrastructure for smoke testing
* add a simple package for basic integration smoke testing

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
log: Overhaul Windows Logger with Pipes and FHs

Redo how we handle IO direction for logging on Windows

Previous we closed sys.stdout, replaced it with a new buffer pointing
at the logging file, and then dup2'd to redirect stdout.
The logfile was then read from a thread, controlled by the main spack
thread, and said thread forwarded on the logging context to main Spack's
dup'd stdout.

This is deeply insufficient for a number of use cases:

It did not support filtering, unicode, control characters, or force_echo

All echo, including force_echo if it did support, was racey, due to the
disparate mechanism for which the reader thread had to control thread behavior
vs where it was obtaining information. I.e. the reader thread could be
killed or echo turned off from pyhon, before any stdout buffers could be flushed
and the reader thread wouldn't see the log output it was supposed to echo before
turning echo off or terminating.

Nested logging, or cached references to stdout and stderr were invalidated
Previously, on io redirection, we were closing stdout and replacing it with
our new buffer. On io restoration, we were restored a buffer pointing at console stdout
to sys.stdout, but any cached references to stdout (i.e. nested logging contexts or pytest)
were invalidated.

Now we:

* do not close whatever sys.stdout refers to, we mearly store a reference to it
* do not construct a new buffer from python's api, instead we perform redirection from
the ground up, starting with fhs to fds and compose a buffer from the existing fh's fd.
* Use pipes for communication with our reader thread, which enables filtering, control characters, etc
and allows for non racey echo behavior.

At a glance:

* Redirects IO starting at the Windows file handle (fh) level
* Properly maps fhs to fds, and obtains the appropriate buffer wrapping each
* Enables proper subprocess io redirection inheritance
* Manages shared file handles between stderr and stdout on redirect
* Properly redirects IO at each level of handling, kernel, C, and python
* Ensures cached references to console streams are not invalidated
* Ensures correct order of release for buffers -> fhs

Moves toward a unified logging class by introducing fundamental support
 for the concepts leveraged by the non Windows logger classes like pipes

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Can't annotate type only avail on Windows

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
No dup_fh on non Windows

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
filter_fn is optional

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
windows only method

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
How did this pass style?

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Import fixes

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
sigh, mypy

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
io isolation in new logging test

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
wip

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Quick review updates

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Loosen Checks for MacOS in CI

MacOS GH CI tests have some issue with coverage not finding the "bin" module that also does not occur with local testing on mac, linux, or Windows. Loosen the check so the extra error cruft of converage does not cause the test to fail, and we still check what we need to check

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Small updates

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Log write only

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Document line ending test behavior

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
use cstdlib headers

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
use cstdlib headers

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
refactor integration yaml workflow

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
refactor integration yaml workflow

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Windows Shells: Allow bin/spack to work as expected

Previously scripts assumed you were working in an established spack shell. This allows them to be invoked directly like bin/spack (or bin\spack for cmd)

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
sha for artifact action

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
sha for artifact action

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
windows win32

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
windows win32

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
better shell

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
better shell

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
more formatting

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
more formatting

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
try this

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
try this

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
use abspath

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
use abspath

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
duh

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
duh

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
use templated var instead of env

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
use templated var instead of env

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
test

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
test

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Restore linux && mac integration runs

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
No gh var expansion

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
try develop

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
develop needs -p

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Use setup script

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
style

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
lol --name

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Add documentation for adding future integration tests

Co-authored-by: Peter Scheibel <scheibel1@llnl.gov>

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Add simple download integration test

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Add test instruction to readme

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
style

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Actually test new package

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Run integration tests like all others

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
all needs integration

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
un-reverse arg order

Signed-off-by: John Parent <john.parent@kitware.com>
